Forward indicators to STIX-Shifter (#118)
tenzir/threatbus: apps/stix-shifter/README.md

STIX-Shifter Threat Bus
=======================

This app bridges the gap between Threat Bus and commercial security tools by
leveraging
[STIX-Shifter](https://github.com/opencybersecurityalliance/stix-shifter).

STIX-Shifter is a tool and library to transform STIX patterns into native
queries for a variety of (mostly) commercial tools, like
[IBM QRadar](https://www.ibm.com/security/security-intelligence/qradar) or

> [Splunk](https://www.splunk.com/). Since Threat Bus itself primarily focuses on
> open-source tools, this app now provides a simple way to use intelligence from
> Threat Bus with the commercial tools of your choice.

Reviewer: ```suggestion
[Splunk](https://www.splunk.com/). This app provides a simple way to communicate STIX
Threat Bus with the commercial tools of your choice.
```

As user, I probably run a mix of commercial and open-source tools, so this distinction does not feel relevant.

## How It Works

The `stix-shifter-threatbus` app uses ZeroMQ to connect with Threat Bus. To
connect via ZeroMQ, users must first install and configure the
[`threatbus-zmq-app`](https://pypi.org/project/threatbus-zmq-app/) plugin on
their Threat Bus host.

> This app functions as middleman between Threat Bus and commercial security
> products. It subscribes to indicator updates from the bus and uses STIX-Shifter

Reviewer: ```suggestion
This app functions as middleman between Threat Bus and security tools support by STIX-Shifter. It subscribes to indicator updates from the bus and uses STIX-Shifter
```

to actively translate STIX-2 intelligence to native queries for commercial
tools. The app then executes these queries via STIX-Shifter. [Result processing

> is yet to be implemented.]

Reviewer: As a reader, I wonder what's happening with the results? Aren't those the retro sightings?

Author: Yes, but that is not yet implemented and will be part of another story (ch24866)

## Quick Start

You can configure the app via a YAML configuration file. See
`config.yaml.example` for an example config file.

Install `stix-shifter-threatbus` in a virtualenv and start it by passing a
config file:

```sh
python -m venv venv
source venv/bin/activate
make dev-mode
stix-shifter-threatbus -c config.yaml
```

## Configuration

Apart from the logging section, which is self-explanatory, users need to
configure the `threatbus` endpoint of the ZerMQ-App plugin and an optional
`snapshot` of historic threat intel data they want to fetch.

Additionally, users must configure each STIX-Shifter module individually to use
it with this app. You also must install the corresponding modules according to
your configuration. For example, if you configure a key `splunk` in the
`modules` section, you must install the `stix-shifter-modules-splunk`. Otherwise

> the app will throw an error. See below for an example:

Author: I'd argue the other way round: usually, when somebody configures a new program they read the docs. For me, the `configuration` section is the right place to describe if a certain config entry requires installing a thrid-party tool, specially if you reference that said tool in your config.

> the app will throw an error. See below for an example:

Reviewer: I still see conditional installation of dependencies as part of the install step. I think then we have to disagree on this. 🙂

```
threatbus: localhost:13370 # connect with Threat Bus via this endpoint
snapshot: 300 # request 300 days of historic indicators
modules:
  # for details on a module's options, please see https://github.com/opencybersecurityalliance/stix-shifter/blob/master/OVERVIEW.md#how-to-use
  # to use the key `splunk` you must install `stix-shifter-modules-splunk`
  # same goes for any other key, e.g., `elastic`, `qradar`, etc...
  splunk:
    max_results: 100 # limit the number of events queried by STIX-Shifter
    # https://github.com/opencybersecurityalliance/stix-shifter/blob/master/OVERVIEW.md#connection
    connection:
      host: localhost
      port: 8089
      selfSignedCert: false
    # https://github.com/opencybersecurityalliance/stix-shifter/blob/master/OVERVIEW.md#configuration
    transmission:
      auth:
        username: admin
        password: admin123
    # https://github.com/opencybersecurityalliance/stix-shifter/blob/master/OVERVIEW.md#translate

>     tranlation: {<Any required options specific to the particular data source>}

Reviewer: Ideally, the example is valid.

```suggestion
    translation: {<Any required options specific to the particular data source>}
```
